mountpoint-s3/tests/fuse_tests/write_test.rs

```diff
@@ -79,70 +93,59 @@ where
+        assert_eq!(err.kind(), ErrorKind::PermissionDenied);
+    } else {
+        // Now test file overwrites
+        let mut f = open_for_write(&path, append, readwrite).expect("should be able to open the file");
+        let mut body = vec![0u8; OBJECT_SIZE];
+        rng.fill(&mut body[..]);
 
-    let mut f = File::options().read(true).write(true).create(true).open(&path).unwrap();
+        let mut current_size = 0;
+        for part in body.chunks(WRITE_SIZE) {
+            let result = f.write_all(part);
 
-    // The file is visible with size 0 as soon as we open it for write
-    let m = metadata(&path).unwrap();
-    assert_eq!(m.len(), 0);
+            assert!(result.is_ok());
+            current_size += part.len() as u64;
+            assert_eq!(f.metadata().unwrap().len(), current_size);
+        }
 
-    // verify the new file is visible in readdir
-    let read_dir_iter = read_dir(&subdir).unwrap();
-    let dir_entry_names = read_dir_to_entry_names(read_dir_iter);
-    assert_eq!(dir_entry_names, vec!["hello.txt", "new.txt", "new_rdwr.txt"]);
+        f.sync_all().unwrap();
 
-    let mut rng = ChaCha20Rng::seed_from_u64(0x12345678 + OBJECT_SIZE as u64);
-    let mut body = vec![0u8; OBJECT_SIZE];
-    rng.fill(&mut body[..]);
+        let m = metadata(&path).unwrap();
+        assert_eq!(m.len(), body.len() as u64);
 
-    for part in body.chunks(WRITE_SIZE) {
-        f.write_all(part).unwrap();
+        let buf = read(&path).unwrap();
+        assert_eq!(&buf[..], &body[..]);
     }
-
-    // We shouldn't be able to read from a file mid-write
-    f.seek(std::io::SeekFrom::End(-1)).unwrap();
-    let err = f.read(&mut [0u8; 1]).expect_err("can't read file while writing");
-    assert_eq!(err.raw_os_error(), Some(libc::EBADF));
-
-    f.sync_all().unwrap();
-    drop(f);
-
-    // Now it's closed, we can stat or read it
-    let m = metadata(&path).unwrap();
-    assert_eq!(m.len(), body.len() as u64);
-
-    let buf = read(&path).unwrap();
-    assert_eq!(&buf[..], &body[..]);
-
-    // Readdir should still work correctly
-    let read_dir_iter = read_dir(&subdir).unwrap();
-    let dir_entry_names = read_dir_to_entry_names(read_dir_iter);
-    assert_eq!(dir_entry_names, vec!["hello.txt", "new.txt", "new_rdwr.txt"]);
-
-    // We shouldn't be allowed to open the file for writing again
-    let err = open_for_write(&path, append).expect_err("can't write existing file");
-    assert_eq!(err.kind(), ErrorKind::PermissionDenied);
 }
 
 #[cfg(feature = "s3_tests")]
-#[test_case(true; "append")]
-#[test_case(false; "no append")]
-fn sequential_write_test_s3(append: bool) {
-    sequential_write_test(crate::fuse_tests::s3_session::new, "sequential_write_test", append);
+#[test_case(true, false; "append write only")]
+#[test_case(true, true; "append readwrite")]
+#[test_case(false, false; "no append write only")]
+#[test_case(false, true; "no append readwrite")]
+fn sequential_write_test_s3(append: bool, readwrite: bool) {
```

Comment:
Do we have a test somewhere that tries to open a file for writing while it's currently open for reading?